internal/guts_cli/guts_cli.go

```diff
@@ -0,0 +1,314 @@
+package guts_cli
+
+// Low level access to pages / data-structures of the bbolt file.
+
+// TODO(ptab): Merge with bbolt/page file that should get ported to internal.
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"unsafe"
+)
+
+var (
+	// ErrCorrupt is returned when a checking a data file finds errors.
+	ErrCorrupt = errors.New("invalid value")
+)
+
+// PageHeaderSize represents the size of the bolt.Page header.
+const PageHeaderSize = 16
+
+// Represents a marker value to indicate that a file (Meta Page) is a Bolt DB.
+const magic uint32 = 0xED0CDAED
+
+// DO NOT EDIT. Copied from the "bolt" package.
+const maxAllocSize = 0xFFFFFFF
+
+// DO NOT EDIT. Copied from the "bolt" package.
+const (
+	branchPageFlag   = 0x01
+	leafPageFlag     = 0x02
+	metaPageFlag     = 0x04
+	freelistPageFlag = 0x10
+)
+
+// DO NOT EDIT. Copied from the "bolt" package.
+const bucketLeafFlag = 0x01
+
+// DO NOT EDIT. Copied from the "bolt" package.
+type Pgid uint64
+
+// DO NOT EDIT. Copied from the "bolt" package.
+type txid uint64
+
+// DO NOT EDIT. Copied from the "bolt" package.
+type Meta struct {
+	magic    uint32
+	version  uint32
+	pageSize uint32
+	flags    uint32
+	root     Bucket
+	freelist Pgid
+	pgid     Pgid // High Water Mark (id of next added Page if the file growths)
+	txid     txid
+	checksum uint64
+}
+
+func LoadPageMeta(buf []byte) *Meta {
+	return (*Meta)(unsafe.Pointer(&buf[PageHeaderSize]))
+}
+
+func (m *Meta) RootBucket() *Bucket {
```

Comment:
It's dangerous to expose the root bucket to users. In theory, users can write k/v pairs to root bucket directly. But since there is no way for users to get the root bucket, so users have to create buckets firstly, and then write k/v pairs to the created buckets. See one example [bucketDemo/main.go#L73-L85](https://github.com/ahrtr/etcd-issues/blob/d0efef10e0fe48757f5bdc27816bc09146ebf25c/bolt/bucketDemo/main.go#L73-L85
)

It's similar to Relational DB, users can't write records to db directly. Instead, they must create tables firstly, and then write records to the created tables. 

So I suggest not to expose the root bucket, so as to keep the behavior/design, and not to break anything.

Reply:
> We expose it only to the bbolt CLI implementation.

If `bbolt CLI implementation` can use the root bucket directly, why can't a user's program/CLI implementation to use it?

Reply:
Got it, the internal package' (https://go.dev/doc/go1.4#internalpackages) prevents it from being imported by packages outside the source subtree.

Reply:
Yes. We ewentually could release public emergency API... but we are far from it.